FRA Cat4 Integration

elipe17 opened the pull request:
## Summary of Changes
- Updated existing FRA validators to be classified as category4 validators
- Added DuplicateManager as a member to FRA to track dups
- Added duplicate detection logic to FRA parse_and_validate
- Moved TDR hash function members to base schema class
- Updated tests to handle duplicates

Pull request closes #3522 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Submit FRA files
3. Verify all errors (except for the "no records generated" error) are cat4
4. Verify exact duplicate records generate an error and are not serialized to the DB
5. Verify the status of all FRA files is Accepted if no errors, Partially Accepted, or Rejected in the event no records were serialized.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] [**_insert ACs here_**]
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Create inner _delete_serialized_records since it is relevant for FRA now too
- Dup manager now sends all fields to error generator
- Add DuplicateManager as member to FRA parser
- Move hash functions to base RowSchema
- Update FRA row schema to generate all cat4 errors
- Update test_parse_fra_ofa_test_cases to test records are removed if serialized when case has cat4 error. Also accounts for duplicate checks
- Updated test files to include more dup cases

elipe17 committed:
- Update test_parse_fra_work_outcome_exiters to correct parms

elipe17 committed:
- Update api tests to use correct year and quarter. Also added more cell checks
- Update cat4 errors to impute empty string for item and column in parser errors

elipe17 committed:
- Update logging in fra schema to be more understandable

elipe17 committed:
- linting

elipe17 committed:
- Add total aggregates support for FRA

elipe17 committed:
Merge branch '3419-fra-error-report' of https://github.com/raft-tech/TANF-app into 3521-fra-cat4-implementation

elipe17 committed:
Merge branch '3419-fra-error-report' of https://github.com/raft-tech/TANF-app into 3521-fra-cat4-implementation

elipe17 committed:
Merge branch '3419-fra-error-report' of https://github.com/raft-tech/TANF-app into 3521-fra-cat4-implementation

jtimpe reviewed (approved):
Looks good!

mattcoleanderson reviewed (approved):
LGTM

raftmsohani reviewed (approved):
LGTM!

elipe17 committed:
Merge branch '3419-fra-error-report' of https://github.com/raft-tech/TANF-app into 3521-fra-cat4-implementation

elipe17 committed:
- Update tests

ADPennington commented:
@elipe17 please check out merge conflicts here when you have time

elipe17 committed:
Merge branch '3419-fra-error-report' of https://github.com/raft-tech/TANF-app into 3521-fra-cat4-implementation

elipe17 committed:
- resolve merge conflicts and update document refs to model

lhuxraft commented:
> @elipe17 please check out merge conflicts here when you have time

@ADPennington Unblocked now - removing the Blocked label

ADPennington commented:
below are my testing notes.

@elipe17 let me know if any of the results captured below are unexpected. @klinkoberstar let me know if this what you had in mind re: not overwhelming users with multiple errors per row. I'm a bit concerned about users not getting the full scope of error feedback, especially because our team will also not have visibility into the full range of errors, but this could be explored through early access research. in the meantime, it seems like the expectation is that once one error in the row is resolved, the file will be resubmitted and the the system will return the other error in the row.  cc: @lhuxraft 

---

-  FRA error feedback will be limited to one error per row. the sequence of checks appears to be the following: 

  >1. check that the exit date is aligned with the fiscal quarter selections
  >2. check that the SSN is valid
  >3. check for duplicate records

- if exit date is invalid, return an error and don't validate SSN (which could also be invalid)

> e.g. Exit date (202508) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
- if exit date is valid, check SSN. if SSN is invalid, return an error
> e.g. Social Security Number is not valid. Check that the Social Security Number is 9 digits, does not contain only zeroes in any one section, and does not contain dashes or other punctuation. Enter 999999999 if an individual does not have an Social Security Number.
- if duplicate record with valid exit or SSN, then the error report will capture the dup error
> e.g. Duplicate record detected with record type TE1 at line 16. Record is a duplicate of the record at line number 15.
- if duplicate record with invalid exit or SSN, then the error report will capture the invalid exit or SSN error (again, if both are invalid, just the exit date will be flagged)
> e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)

elipe17 commented:
> below are my testing notes.
> 
> @elipe17 let me know if any of the results captured below are unexpected. @klinkoberstar let me know if this what you had in mind re: not overwhelming users with multiple errors per row. I'm a bit concerned about users not getting the full scope of error feedback, especially because our team will also not have visibility into the full range of errors, but this could be explored through early access research. in the meantime, it seems like the expectation is that once one error in the row is resolved, the file will be resubmitted and the the system will return the other error in the row. cc: @lhuxraft
> 
> * FRA error feedback will be limited to one error per row. the sequence of checks appears to be the following:
> 
> > 1. check that the exit date is aligned with the fiscal quarter selections
> > 2. check that the SSN is valid
> > 3. check for duplicate records
> 
> * if exit date is invalid, return an error and don't validate SSN (which could also be invalid)
> 
> > e.g. Exit date (202508) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> 
> * if exit date is valid, check SSN. if SSN is invalid, return an error
> 
> > e.g. Social Security Number is not valid. Check that the Social Security Number is 9 digits, does not contain only zeroes in any one section, and does not contain dashes or other punctuation. Enter 999999999 if an individual does not have an Social Security Number.
> 
> * if duplicate record with valid exit or SSN, then the error report will capture the dup error
> 
> > e.g. Duplicate record detected with record type TE1 at line 16. Record is a duplicate of the record at line number 15.
> 
> * if duplicate record with invalid exit or SSN, then the error report will capture the invalid exit or SSN error (again, if both are invalid, just the exit date will be flagged)
> 
> > e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> > e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)

Yep I see the issue. Giving the user a more complete picture of the errors will help prevent extra submissions. I should be able to all all the errors to fall through so that we see everything before deciding to serialize the record or not. Let me know if that is your preference and I can make the changes to support that.

klinkoberstar commented:
> > below are my testing notes.
> > @elipe17 let me know if any of the results captured below are unexpected. @klinkoberstar let me know if this what you had in mind re: not overwhelming users with multiple errors per row. I'm a bit concerned about users not getting the full scope of error feedback, especially because our team will also not have visibility into the full range of errors, but this could be explored through early access research. in the meantime, it seems like the expectation is that once one error in the row is resolved, the file will be resubmitted and the the system will return the other error in the row. cc: @lhuxraft
> > 
> > * FRA error feedback will be limited to one error per row. the sequence of checks appears to be the following:
> > 
> > > 1. check that the exit date is aligned with the fiscal quarter selections
> > > 2. check that the SSN is valid
> > > 3. check for duplicate records
> > 
> > 
> > 
> > * if exit date is invalid, return an error and don't validate SSN (which could also be invalid)
> > 
> > > e.g. Exit date (202508) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> > 
> > 
> > 
> > * if exit date is valid, check SSN. if SSN is invalid, return an error
> > 
> > > e.g. Social Security Number is not valid. Check that the Social Security Number is 9 digits, does not contain only zeroes in any one section, and does not contain dashes or other punctuation. Enter 999999999 if an individual does not have an Social Security Number.
> > 
> > 
> > 
> > * if duplicate record with valid exit or SSN, then the error report will capture the dup error
> > 
> > > e.g. Duplicate record detected with record type TE1 at line 16. Record is a duplicate of the record at line number 15.
> > 
> > 
> > 
> > * if duplicate record with invalid exit or SSN, then the error report will capture the invalid exit or SSN error (again, if both are invalid, just the exit date will be flagged)
> > 
> > > e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> > > e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> 
> Yep I see the issue. Giving the user a more complete picture of the errors will help prevent extra submissions. I should be able to all all the errors to fall through so that we see everything before deciding to serialize the record or not. Let me know if that is your preference and I can make the changes to support that.

I envisioned something like one error per column per row: if the date and SSN had issues, 2 errors; if the SSN violated multiple conditions, we would only return one SSN error. 

We did discuss in DIGIT that we want to be strict about the validation rules in this first pass to get people to adopt the file format guidance early and not let anything through that isn't valid. To that end, this approach is kind of appealing. But agree that for certain cases it leaves them without enough info to correct.

Ultimately, I think we're operating on some assumptions about how easy it will or won't be for users to configure this file and make corrections. I'm not sure if this will be too little or just enough. Regardless of how we handle this now, we should definitely explore this in early access testing. 

I think this is on UX's radar, but tagging for visibility
cc: @reitermb @victoriaatraft

elipe17 commented:
> > > below are my testing notes.
> > > @elipe17 let me know if any of the results captured below are unexpected. @klinkoberstar let me know if this what you had in mind re: not overwhelming users with multiple errors per row. I'm a bit concerned about users not getting the full scope of error feedback, especially because our team will also not have visibility into the full range of errors, but this could be explored through early access research. in the meantime, it seems like the expectation is that once one error in the row is resolved, the file will be resubmitted and the the system will return the other error in the row. cc: @lhuxraft
> > > 
> > > * FRA error feedback will be limited to one error per row. the sequence of checks appears to be the following:
> > > 
> > > > 1. check that the exit date is aligned with the fiscal quarter selections
> > > > 2. check that the SSN is valid
> > > > 3. check for duplicate records
> > > 
> > > 
> > > 
> > > * if exit date is invalid, return an error and don't validate SSN (which could also be invalid)
> > > 
> > > > e.g. Exit date (202508) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> > > 
> > > 
> > > 
> > > * if exit date is valid, check SSN. if SSN is invalid, return an error
> > > 
> > > > e.g. Social Security Number is not valid. Check that the Social Security Number is 9 digits, does not contain only zeroes in any one section, and does not contain dashes or other punctuation. Enter 999999999 if an individual does not have an Social Security Number.
> > > 
> > > 
> > > 
> > > * if duplicate record with valid exit or SSN, then the error report will capture the dup error
> > > 
> > > > e.g. Duplicate record detected with record type TE1 at line 16. Record is a duplicate of the record at line number 15.
> > > 
> > > 
> > > 
> > > * if duplicate record with invalid exit or SSN, then the error report will capture the invalid exit or SSN error (again, if both are invalid, just the exit date will be flagged)
> > > 
> > > > e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> > > > e.g. Exit date (202604) is not valid. Date must be in the range of 2025 - Q3 (Apr - Jun)
> > 
> > 
> > Yep I see the issue. Giving the user a more complete picture of the errors will help prevent extra submissions. I should be able to all all the errors to fall through so that we see everything before deciding to serialize the record or not. Let me know if that is your preference and I can make the changes to support that.
> 
> I envisioned something like one error per column per row: if the date and SSN had issues, 2 errors; if the SSN violated multiple conditions, we would only return one SSN error.
> 
> We did discuss in DIGIT that we want to be strict about the validation rules in this first pass to get people to adopt the file format guidance early and not let anything through that isn't valid. To that end, this approach is kind of appealing. But agree that for certain cases it leaves them without enough info to correct.
> 
> Ultimately, I think we're operating on some assumptions about how easy it will or won't be for users to configure this file and make corrections. I'm not sure if this will be too little or just enough. Regardless of how we handle this now, we should definitely explore this in early access testing.
> 
> I think this is on UX's radar, but tagging for visibility cc: @reitermb @victoriaatraft

I can go ahead and make the changes to be a bit more verbose.

ADPennington commented:
> I can go ahead and make the changes to be a bit more verbose.

as per today's digit discussion, there is a shorter turnaround for these reports since we need to match quickly with external data source given ndnh data retention policies...so one error per field and row is optimal here. @elipe17 cc: @lhuxraft @reitermb @victoriaatraft @klinkoberstar

elipe17 committed:
- Updated fra parser to have fall through validation
- Updated tests to match expectations

elipe17 commented:
> Collaborator

@ADPennington & @klinkoberstar, I pushed up changes which allows all validation to happen (and therefore error generation) before deciding whether to exclude a record which would in turn prevent further validation. There should now be at most three errors generated per row in an FRA submission: EXIT_DATE issue, SSN issue, or a duplicate detected.

ADPennington commented:
@elipe17 exact dups are okay when SSN is all 9s, since all 9s are allowable. currently triggering a dup 

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/APENNI~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/APENNI~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{font-weight:700;}
-->
</style>
</head>

<body link=blue vlink=purple>


Exit Date | SSN | Row Number In File | Error Description
-- | -- | -- | --
202506 | *****9999 | 4 | Duplicate record detected with record type TE1 at line 4. Record is a   duplicate of the record at line number 3.



</body>

</html>

elipe17 commented:
> @elipe17 exact dups are okay when SSN is all 9s, since all 9s are allowable. currently triggering a dup
> 
> <style> </style>
> Exit Date	SSN	Row Number In File	Error Description
> 202506	*****9999	4	Duplicate record detected with record type TE1 at line 4. Record is a duplicate of the record at line number 3.

@ADPennington is this a blocking change? I can make it so that the duplicate detection is cognizant of those types of things.  But it would probably take me about a day to get that coded up and tested.

lhuxraft commented:
> > @elipe17 exact dups are okay when SSN is all 9s, since all 9s are allowable. currently triggering a dup
> > <style> </style>
> > Exit Date	SSN	Row Number In File	Error Description
> > 202506	*****9999	4	Duplicate record detected with record type TE1 at line 4. Record is a duplicate of the record at line number 3.
> 
> @ADPennington is this a blocking change? I can make it so that the duplicate detection is cognizant of those types of things. But it would probably take me about a day to get that coded up and tested.

@elipe17 This should be blocking. There were a lot of questions from users on this last week, so if we're able to go ahead and get it implemented, that would be great 😄 This, and making the duplicate error message language consistent with the mockup, should be the last changes for this to be GTG though

elipe17 committed:
- Created new base class for duplicate detectors
- Add factory class for duplicate manager to use to determine what detector to instantiate
- Updated TE1 error message to align with expectations

elipe17 committed:
- Update TE1 dup detector to ignore records with SSN === all 9's
- Added test case to test files

elipe17 committed:
- linting

elipe17 committed:
Merge branch 'develop' into 3521-fra-cat4-implementation

elipe17 commented:
> > > @elipe17 exact dups are okay when SSN is all 9s, since all 9s are allowable. currently triggering a dup
> > > <style> </style>
> > > Exit Date	SSN	Row Number In File	Error Description
> > > 202506	*****9999	4	Duplicate record detected with record type TE1 at line 4. Record is a duplicate of the record at line number 3.
> > 
> > 
> > @ADPennington is this a blocking change? I can make it so that the duplicate detection is cognizant of those types of things. But it would probably take me about a day to get that coded up and tested.
> 
> @elipe17 This should be blocking. There were a lot of questions from users on this last week, so if we're able to go ahead and get it implemented, that would be great 😄 This, and making the duplicate error message language consistent with the mockup, should be the last changes for this to be GTG though

@ADPennington & @lhuxraft, I pushed changes that aligns the error message with given message. I did add to it the location the submitter should look in the file to verify the duplicate. I also added the logic to not apply duplicate detection if the SSN is all 9's.

elipe17 committed:
- grammar fix

elipe17 committed:
Merge branch '3521-fra-cat4-implementation' of https://github.com/raft-tech/TANF-app into 3521-fra-cat4-implementation

ADPennington commented:
@lhuxraft @reitermb @victoriaatraft @klinkoberstar 

below is a test file and associated error report as of this PR. FYSA. 

[FY2025Q1_TE_Rpt_NY.csv](https://github.com/user-attachments/files/19649771/FY2025Q1_TE_Rpt_NY.csv)
[2025-Q1-Work Outcomes of TANF Exiters (3).xlsx](https://github.com/user-attachments/files/19649776/2025-Q1-Work.Outcomes.of.TANF.Exiters.3.xlsx)

ADPennington commented on tdrs-backend/tdpservice/parsers/duplicate_detectors.py:
```diff
@@ -0,0 +1,238 @@
+        # created from it will have all of it's records appropriately marked for deletion if need be.
+        ssn = getattr(record, 'SSN', None)
+        if ssn == "999999999":
+            logger.warning(f"SSN is 999999999 for record {record.id} at line number "
```
very nice.

ADPennington reviewed (approved):
@elipe17 LGTM 🚀 updated testing notes below:

cc: @klinkoberstar @lhuxraft @reitermb 

---

- FRA error feedback captures up to 3 types of generic errors per row: 

  >1. check that the exit date is valid which means:
  >>-  aligned with the fiscal quarter selections
  >>- year >= 2024
  >>- month >= 01 and <= 12
  
  >2. check that the SSN is "valid" which means:
  >>- 9 digits, 
  >>- does not contain only zeroes in any one section, 
  >>- does not contain dashes or other punctuation. 
  
  >3. check for duplicate records

- if exit date is invalid, error message looks like the following:

>>`Exit date (202413) is not valid. Date must be in the range of 2025 - Q1 (Oct - Dec)`


- if SSN is invalid, error message looks like the following:

>> `Social Security Number is not valid. Check that the Social Security Number is 9 digits, does not contain only zeroes in any one section, and does not contain dashes or other punctuation. Enter 999999999 if an individual does not have an Social Security Number.`

- if duplicate record, error message looks like the following: 

>> `Duplicate Social Security Number within a month. Check that individual Social Security Numbers within a single exit month are not included more than once. Social Security Number is a duplicate of the record at line number 19.`

- Error report changes (see below)
![Screenshot 2025-04-08 114914](https://github.com/user-attachments/assets/eca3013d-fd43-4e0e-b91e-29af1b8b818d)

elipe17 committed:
Merge branch 'develop' into 3521-fra-cat4-implementation

elipe17 committed:
Merge branch 'develop' into 3521-fra-cat4-implementation

elipe17 committed:
Merge branch 'develop' into 3521-fra-cat4-implementation

elipe17 committed:
Merge branch 'develop' into 3521-fra-cat4-implementation

elipe17 committed:
Merge branch 'develop' into 3521-fra-cat4-implementation
